### DISCUSSION TITLE
Finding USGS Water Quality Data w/ ChatGPT

### DISCUSSION THREAD

**Discussion by nquarder** [cu-esiil-edu/stars-2024#34] [Q&A]
@tes-ani asked about finding USGS water quality data for the Colorado River above Glenwood Springs, specifically looking at turbidity (water cloudiness). USGS Waterdata is a good resource and it can be tricky to access and open data outside of their dashboard. 

Here is a link to the site with a plot of turbidity for the last 7 days: https://waterdata.usgs.gov/monitoring-location/09071750/#parameterCode=63680&period=P7D&showMedian=true

From here it looks like turbidity data goes back to October 2020, a couple of months after the Grizzly Creek Fire event that took place just above this river gage. 

Hydrofunctions is a python library that allows you to pull in river hydrology data from the USGS National Weather Information System: https://hydrofunctions.readthedocs.io/en/master/introduction.html#features

We have some lessons in our open EDS textbook that teaches users how to work with hydrofunctions: 
* https://www.earthdatascience.org/courses/use-data-open-source-python/use-time-series-data-in-python/date-time-types-in-pandas-python/time-series-exercise/
* https://www.earthdatascience.org/courses/use-data-open-source-python/use-time-series-data-in-python/floods-return-period-and-probability/

I [asked ChatGPT](https://chatgpt.com/c/5800018c-d1e5-41d6-b79d-c059e17ef173) to write some code to access the turbidity data from this particular river gage and here is what it wrote for me after some back and forth:

```
import hydrofunctions as hf
import matplotlib.pyplot as plt

# Ensure plots are displayed inline in the notebook
%matplotlib inline

# Define the station ID and the parameter code for turbidity
station_id = '09071750'
parameter_code = '63680'  # The parameter code for turbidity

# Define the start and end dates
start_date = '2020-01-01'
end_date = '2023-12-31'  # Update this to the current date if needed

# Make the API call to get the data
turbidity_data = hf.NWIS(station_id, service='iv', start_date=start_date, end_date=end_date, parameterCd=parameter_code)

# Convert the data to a DataFrame
turbidity_df = turbidity_data.df()

# Display the data
print(turbidity_df)

# Plot the data
plt.figure(figsize=(10, 6))
plt.plot(turbidity_df.index, turbidity_df['USGS:09071750:63680:00000'], label='Turbidity')
plt.xlabel('Date')
plt.ylabel('Turbidity (NTU)')
plt.title('Turbidity Data for Colorado River above Glenwood Springs')
plt.legend()
plt.grid(True)

# Save the plot as a PNG file
plt.savefig('turbidity_plot3.png')

# Show the plot
plt.show()
```

![image](https://github.com/cu-esiil-edu/stars-2024/assets/16786791/9141188a-e9f6-4a37-b1df-eb4cbdd494bd)

It took some time to get things the way I wanted and a couple rounds trying different things with ChatGPT to get what I was looking for (apologies for deleting the original conversation!). For folks that are looking for data for your projects, you might give something like this a try. We are happy to help demo something like this too during office hours or in another meeting. -NQ

**Comment by tes-ani** [comment]
Thanks so much, Nate. This looks awesome and I'm going to play around with this code for our project using a couple other stations as well.
